src/hotspot/share/memory/arena.cpp

```diff
@@ -212,19 +181,37 @@ void Chunk::start_chunk_pool_cleaner_task() {
   cleaner->enroll();
 }
 
-//------------------------------Arena------------------------------------------
+Chunk::Chunk(size_t length) : _len(length) {
+  _next = nullptr;         // Chain on the linked list
+}
+
+void Chunk::chop(Chunk* k) {
+  while (k != nullptr) {
+    Chunk *tmp = k->next();
+    // clear out this chunk (to detect allocation bugs)
+    if (ZapResourceArea) memset(k->bottom(), badResourceValue, k->length());
+    ChunkPool::deallocate_chunk(k);
+    k = tmp;
+  }
+}
+
+void Chunk::next_chop(Chunk* k) {
+  assert(k != nullptr && k->_next != nullptr, "must be non-null");
+  Chunk::chop(k->_next);
+  k->_next = nullptr;
+}
 
 Arena::Arena(MEMFLAGS flag, size_t init_size) : _flags(flag), _size_in_bytes(0)  {
   init_size = ARENA_ALIGN(init_size);
-  _first = _chunk = new (AllocFailStrategy::EXIT_OOM, init_size) Chunk(init_size);
+  _first = _chunk = ChunkPool::allocate_chunk(AllocFailStrategy::EXIT_OOM, init_size);
   _hwm = _chunk->bottom();      // Save the cached hwm, max
   _max = _chunk->top();
   MemTracker::record_new_arena(flag);
   set_size_in_bytes(init_size);
 }
 
 Arena::Arena(MEMFLAGS flag) : _flags(flag), _size_in_bytes(0) {
-  _first = _chunk = new (AllocFailStrategy::EXIT_OOM, Chunk::init_size) Chunk(Chunk::init_size);
+  _first = _chunk = ChunkPool::allocate_chunk(AllocFailStrategy::EXIT_OOM, Chunk::init_size);
```

Comment:
I have one request - can you make the AllocFailStrategy the last parameter?

Reply:
Fixed!